internal/pkg/v2/ui/http/handle/usecase.go

```diff
@@ -0,0 +1,110 @@
+/*******************************************************************************
+ * Copyright 2020 Dell Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ *******************************************************************************/
+
+package handle
+
+import (
+	"encoding/json"
+	"io"
+	"net/http"
+
+	"github.com/edgexfoundry/edgex-go/internal/pkg/concurrent"
+	"github.com/edgexfoundry/edgex-go/internal/pkg/v2/application"
+	dtoBase "github.com/edgexfoundry/edgex-go/internal/pkg/v2/application/dto/v2dot0/base"
+	dtoError "github.com/edgexfoundry/edgex-go/internal/pkg/v2/application/dto/v2dot0/error"
+	"github.com/edgexfoundry/edgex-go/internal/pkg/v2/ui/http/routing"
+)
+
+// UseCaseRequest decodes, processes, and returns a result for a use-case-specific endpoint's request.
+func UseCaseRequest(w http.ResponseWriter, r *http.Request, version, kind, action string, router *routing.Router) {
+	defer func() {
+		if r.Body != nil {
+			_ = r.Body.Close()
+		}
+	}()
+
+	routable, exists := router.FindRoute(version, kind, action)
+	if !exists {
+		http.Error(w, "route not found", http.StatusNotFound)
+		return
+	}
+
+	request := routable.Request()
+	// is request for a use case that doesn't have a request DTO defined (i.e. routable.Request() returns nil)?
+	if request == nil {
+		response, status := routable.Execute(request)
+		httpJSONResult(w, statusToHTTPStatusCode(status), response)
+		return
+	}
+
+	body := make([]byte, r.ContentLength)
+	_, err := r.Body.Read(body)
+	if err != nil && err != io.EOF {
+		httpJSONResult(
+			w,
+			http.StatusBadRequest,
+			dtoError.NewResponse(dtoBase.NewResponse("", err.Error(), application.StatusUseCaseContentErrorFailure)),
+		)
+		return
+	}
+
+	// multiple requests?
+	requests := []*json.RawMessage{}
+	err = json.Unmarshal(body, &requests)
```

Comment:
I wonder if there is a better way to do this. With this layout we are forcing the caller to put all information in the body of the request rather than leveraging the HTTP transport for information, commonly done in REST. I do not think GET requests that just require an ID should contain a HTTP request body, since the information is in the URI. I also think it is important to have an internal representation of the DTO which encapsulates the request information to be processed regardless of transport. My thought is can we add functionality here which allows us to pull information from places other than just the HTTP body so that our APIs are more REST friendly?

Reply:
> I do not think GET requests that just require an ID should contain a HTTP request body

It isn't the intention to have HTTP GET or DELETE requests to require an HTTP request body.  At the moment, the existing implementation doesn't have an example of GET or DELETE HTTP method requests; this is something I will be adding with the follow-on service-specific use-case example issue.

Implementation of GET and DELETE requests will -- within the corresponding controllers -- translate transport-specific implementation (such as identity or other content embedded in the requested URL; a security token passed in an HTTP header; etc.) to and from individual use-case request and response DTOs.

Unresolve the conversation and ask additional questions if this is not an adequate explanation.

Reply:
I phrased my comment very poorly. What I meant was that in situations that only require an ID like getting an entity, which is typically done with a `GET` with the ID in the URI, would have to be be changed to a POST with the ID in the body with this implementation. This is because `GET` does not have a body and therefore we cannot get any context for the request.

With that said, I still believe you addressed my core concern of adding functionality to support data coming in via different HTTP mechanisms such as URI, headers, and query parameters.